File: src/main/java/com/indiepost/repository/jpa/BookmarkRepositoryJpa.java
package com.indiepost.repository.jpa;

import com.indiepost.model.Bookmark;
import com.indiepost.model.Post;
import com.indiepost.model.QBookmark;
import com.indiepost.model.User;
import com.indiepost.repository.BookmarkRepository;
import com.querydsl.jpa.impl.JPAQueryFactory;
import org.springframework.stereotype.Repository;

import javax.persistence.EntityManager;
import javax.persistence.PersistenceContext;
import java.util.List;

@Repository
public class BookmarkRepositoryJpa implements BookmarkRepository {
    @PersistenceContext
    private EntityManager entityManager;

    @Override
    public void save(Bookmark bookmark) {
        entityManager.persist(bookmark);
    }

    @Override
    public void create(Long userId, Long postId) {
        Bookmark bookmark = findOneByUserIdAndPostId(userId, postId);
        if (bookmark != null) {
            return;
        }
        User userRef = entityManager.getReference(User.class, userId);
        Post postRef = entityManager.getReference(Post.class, postId);
        bookmark = new Bookmark(userRef, postRef);
        save(bookmark);
    }

    @Override
    public void delete(Bookmark bookmark) {
        entityManager.remove(bookmark);
    }

    @Override
    public Bookmark findOneByUserIdAndPostId(Long userId, Long postId) {
        QBookmark b = QBookmark.bookmark;
        return getQueryFactory()
                .selectFrom(b)
                .where(b.userId.eq(userId).and(b.postId.eq(postId)))
                .fetchOne();
    }

    @Override
    public List<Bookmark> findByUserIdAndPostIds(Long userId, List<Long> postIds) {
        QBookmark b = QBookmark.bookmark;
        return getQueryFactory()
                .selectFrom(b)
                .where(b.userId.eq(userId).and(b.postId.in(postIds)))
                .fetch();
    }

    @Override
    public void removeAllBookmarksByUserId(Long userId) {
        QBookmark b = QBookmark.bookmark;
        getQueryFactory()
                .delete(b)
                .where(b.userId.eq(userId))
                .execute();
    }

    private JPAQueryFactory getQueryFactory() {
        return new JPAQueryFactory(entityManager);
    }
}

File: src/main/kotlin/com/indiepost/repository/jpa/BookmarkRepositoryJpa.kt
package com.indiepost.repository.jpa

import com.indiepost.model.Bookmark
import com.indiepost.model.Post
import com.indiepost.model.QBookmark
import com.indiepost.model.User
import com.indiepost.repository.BookmarkRepository
import com.querydsl.jpa.impl.JPAQueryFactory
import org.springframework.stereotype.Repository
import java.time.LocalDateTime

import javax.persistence.EntityManager
import javax.persistence.PersistenceContext

@Repository
class BookmarkRepositoryJpa : BookmarkRepository {
    @PersistenceContext
    private lateinit var entityManager: EntityManager

    private val queryFactory: JPAQueryFactory
        get() = JPAQueryFactory(entityManager)

    override fun save(bookmark: Bookmark) {
        entityManager.persist(bookmark)
    }

    override fun create(userId: Long, postId: Long) {
        var bookmark = findOneByUserIdAndPostId(userId, postId)
        if (bookmark != null) {
            return
        }
        val userRef = entityManager.getReference(User::class.java, userId)
        val postRef = entityManager.getReference(Post::class.java, postId)
        bookmark = Bookmark(user = userRef, post = postRef, created = LocalDateTime.now())
        save(bookmark)
    }

    override fun delete(bookmark: Bookmark) {
        entityManager.remove(bookmark)
    }

    override fun findOneByUserIdAndPostId(userId: Long, postId: Long): Bookmark? {
        val b = QBookmark.bookmark
        return queryFactory
                .selectFrom(b)
                .where(b.userId.eq(userId).and(b.postId.eq(postId)))
                .fetchOne()
    }

    override fun findByUserIdAndPostIds(userId: Long, postIds: List<Long>): List<Bookmark> {
        val b = QBookmark.bookmark
        return queryFactory
                .selectFrom(b)
                .where(b.userId.eq(userId).and(b.postId.`in`(postIds)))
                .fetch()
    }

    override fun removeAllBookmarksByUserId(userId: Long) {
        val b = QBookmark.bookmark
        queryFactory
                .delete(b)
                .where(b.userId.eq(userId))
                .execute()
    }
}
